src/wutil/gettext.rs

```diff
@@ -120,25 +117,105 @@ fn wgettext_impl(text: MaybeStatic) -> &'static wstr {
     res
 }
 
-/// Get a (possibly translated) string from a literal.
-/// Note this assumes that the string does not contain interior NUL characters -
-/// this is checked in debug mode.
-pub fn wgettext_static_str(s: &'static wstr) -> &'static wstr {
-    wgettext_impl(MaybeStatic::Static(s))
+/// A string which can be translated.
+/// The wrapped string itself is the original, untranslated version.
+/// Use [`TranslatableString::translate`] to obtain the translation.
+///
+/// Do not construct this type directly.
+/// For string literals defined in fish's Rust sources,
+/// use the macros defined in this file.
+/// For strings defined elsewhere, use [`TranslatableString::from_external_source`].
+/// Use this function with caution. If the string is not extracted into the gettext PO files from
+/// which fish obtains translations, translations will not work.
+#[derive(Debug, Clone)]
+pub enum TranslatableString {
+    Static(&'static wstr),
+    Owned(WString),
+}
+
+impl TranslatableString {
+    /// Create a [`TranslatableString`] from a string which is not from fish's own Rust sources.
+    /// Translation will only work if this string is extracted into the translation files some
+    /// other way.
+    pub fn from_external_source(s: WString) -> Self {
+        Self::Owned(s)
+    }
+
+    /// Get the translation of a [`TranslatableString`].
+    /// If original string is empty, an empty `wstr` is returned,
+    /// instead of the gettext metadata.
+    pub fn translate(&self) -> &'static wstr {
+        match self {
+            Self::Static(s) => {
+                if s.is_empty() {
+                    L!("")
+                } else {
+                    wgettext_impl(MaybeStatic::Static(s))
+                }
+            }
+            Self::Owned(s) => {
+                if s.is_empty() {
+                    L!("")
+                } else {
+                    wgettext_impl(MaybeStatic::Local(s))
+                }
+            }
+        }
+    }
+}
+
+impl std::fmt::Display for TranslatableString {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.translate())
```

Comment:
I think this uses `std::fmt::Write` which can't handle UTF-8 data (we would want to use `wcs2string`).
Looks like this Display implementation is only needed for `FLOG(error, translatable_str)`.
It's fine because we already had that behavior before this change.

(Actually `printf/src/arg.rs` might be broken for us because it does `storage.extend(s.chars());`,
e.g. it skips `wcs2string` which is what we need to decode our PUA chars.)

So yeah, we can clean this up later.